File: src/concurrency/book2/sample/chapter8/InternalTask.java
package concurrency.book2.sample.chapter8;

/**
 * 用于线程池内部，不断从queue中取出某个runnable
 */
public class InternalTask implements Runnable {

    private final RunnableQueue runnableQueue;

    private volatile boolean running = true;

    public InternalTask(RunnableQueue runnableQueue) {
        this.runnableQueue = runnableQueue;
    }

    @Override
    public void run() {

        while (running && !Thread.currentThread().isInterrupted()) {
            try {
                Runnable task = runnableQueue.take();
                task.run();
            } catch (InterruptedException e) {
                running = false;
                break;
            }
        }
    }

    //停止当前任务，主要会在线程的shutdown方法中使用
    public void stop() {
        this.running = false;
    }

}

File: src/concurrency/book2/sample/chapter8/kotlin/InternalTask.kt
//package concurrency.book2.sample.chapter8.kotlin
//
//class InternalTask(private val runnableQueue: RunnableQueue) : Runnable {
//
//
//    @Volatile
//    private var running: Boolean = true
//
//    override fun run() {
//
//        while (running && !Thread.currentThread().isInterrupted) {
//            try {
//                val task = runnableQueue.take()
//                task.run()
//            } catch (e: InterruptedException) {
//                running = false
//                break
//            }
//        }
//    }
//
//    fun stop() {
//        this.running = false
//    }
//}